WeightedSpanTermExtractor is fragile to modern query rewrites due to DelegatingLeafReader limitations

### Description

The `PlainHighlighter` via [WeightedSpanTermExtractor.java#L447](https://github.com/apache/lucene/blob/releases/lucene/10.3.2/lucene/highlighter/src/java/org/apache/lucene/search/highlight/WeightedSpanTermExtractor.java#L447) is prone to crashing when rewriting modern Lucene query types. This happens because the internal `DelegatingLeafReader` used for highlighting throws `UnsupportedOperationException` when `getFieldInfos()` is called. 

**The Recurrent Issue**

As Lucene adds more optimizations to query rewrite() methods (e.g., checking index statistics or DocValues existence), these queries begin to fail when processed by the highlighter.

This pattern was previously observed with `FieldExistsQuery` (fixed in https://github.com/apache/lucene/pull/12088 by explicitly ignoring it). I am now observing the same crash in [OpenSearch](https://github.com/opensearch-project/OpenSearch/issues/20496) with `SortedNumericDocValuesRangeQuery` (via IndexOrDocValuesQuery), which attempts to access DocValuesSkipper and subsequently getFieldInfos() during rewrite

This creates a "whack-a-mole" situation where every new query optimization potentially breaks the highlighter.

**Example Failure**

When highlighting a boolean query containing a range filter:

```
Caused by: java.lang.UnsupportedOperationException
    at org.apache.lucene.search.highlight.WeightedSpanTermExtractor$DelegatingLeafReader.getFieldInfos(WeightedSpanTermExtractor.java:447)
    at org.apache.lucene.index.DocValuesSkipper.globalMinValue(DocValuesSkipper.java:137)
    at org.apache.lucene.document.SortedNumericDocValuesRangeQuery.rewrite(SortedNumericDocValuesRangeQuery.java:101)
```

**Proposal & Discussion**

I believe we should address this structurally rather than adding exceptions for every new query type.

**Option 1**: The Structural Fix - Modify DelegatingLeafReader.getFieldInfos() to return FieldInfos instead of throwing an exception

PS: I am not very familiar with lucene codebase & not so sure if this option is feasible or it may require a long time to fix

**Option 2**: The Targeted Fix - Continue the pattern established in [PR #12088](https://github.com/apache/lucene/pull/12088) by adding IndexOrDocValuesQuery (and others) to the "ignore" list in `WeightedSpanTermExtractor.extract()` method

I am happy to submit a PR for option 2 & maybe option 1 as well if team suggests some solution 

### Version and environment details

_No response_

--- a/lucene/CHANGES.txt
+++ b/lucene/CHANGES.txt
@@ -128,6 +128,9 @@ Bug Fixes
 * GITHUB#12561: UAX29URLEmailTokenizer matched emails with commas and invalid
   periods in the local part. (Eran Yarkon)
 
+* GITHUB#15668: Fix UnsupportedOperationException in WeightedSpanTermExtractor by
+  implementing getFieldInfos in DelegatingLeafReader. (Simranjeet Singh)
+
 Changes in Runtime Behavior
 ---------------------
 * GITHUB#14187: The query cache is now disabled by default. (Adrien Grand)

--- a/lucene/highlighter/src/java/org/apache/lucene/search/highlight/WeightedSpanTermExtractor.java
+++ b/lucene/highlighter/src/java/org/apache/lucene/search/highlight/WeightedSpanTermExtractor.java
@@ -27,6 +27,7 @@ import java.util.Set;
 import org.apache.lucene.analysis.CachingTokenFilter;
 import org.apache.lucene.analysis.TokenStream;
 import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.FieldInfo;
 import org.apache.lucene.index.FieldInfos;
 import org.apache.lucene.index.FilterLeafReader;
 import org.apache.lucene.index.IndexReader;
@@ -444,7 +445,14 @@ public class WeightedSpanTermExtractor {
 
     @Override
     public FieldInfos getFieldInfos() {
-      throw new UnsupportedOperationException(); // TODO merge them
+
+      return new FieldInfos(
+          new FieldInfo[] {super.getFieldInfos().fieldInfo(DelegatingLeafReader.FIELD_NAME)}) {
+        @Override
+        public FieldInfo fieldInfo(String fieldName) {
+          return super.fieldInfo(DelegatingLeafReader.FIELD_NAME);
+        }
+      };
     }
 
     @Override

--- a/lucene/highlighter/src/java/org/apache/lucene/search/highlight/WeightedSpanTermExtractor.java
+++ b/lucene/highlighter/src/java/org/apache/lucene/search/highlight/WeightedSpanTermExtractor.java
@@ -54,7 +54,6 @@ import org.apache.lucene.search.BooleanQuery;
 import org.apache.lucene.search.BoostQuery;
 import org.apache.lucene.search.ConstantScoreQuery;
 import org.apache.lucene.search.DisjunctionMaxQuery;
-import org.apache.lucene.search.FieldExistsQuery;
 import org.apache.lucene.search.IndexOrDocValuesQuery;
 import org.apache.lucene.search.IndexSearcher;
 import org.apache.lucene.search.MatchAllDocsQuery;
@@ -237,8 +236,6 @@ public class WeightedSpanTermExtractor {
         }
       }
     } else if (query instanceof MatchAllDocsQuery) {
-      // nothing
-    } else if (query instanceof FieldExistsQuery) {
       // nothing
     } else if (query instanceof FunctionScoreQuery) {
       extract(((FunctionScoreQuery) query).getWrappedQuery(), boost, terms);

--- a/lucene/highlighter/src/test/org/apache/lucene/search/highlight/TestHighlighter.java
+++ b/lucene/highlighter/src/test/org/apache/lucene/search/highlight/TestHighlighter.java
@@ -41,6 +41,7 @@ import org.apache.lucene.document.Document;
 import org.apache.lucene.document.Field;
 import org.apache.lucene.document.FieldType;
 import org.apache.lucene.document.IntPoint;
+import org.apache.lucene.document.SortedNumericDocValuesField;
 import org.apache.lucene.document.StoredField;
 import org.apache.lucene.document.TextField;
 import org.apache.lucene.index.DirectoryReader;
@@ -2231,6 +2232,23 @@ public class TestHighlighter extends BaseTokenStreamTestCase implements Formatte
       String result = h.getBestFragment(stream, text);
       assertEquals("random <B>words</B> and words", result); // only highlight first "word"
     }
+  }
+
+  public void testSortedNumericDocValuesRangeQuery() throws IOException {
+    Query query =
+        new BooleanQuery.Builder()
+            .add(new TermQuery(new Term("text", "compensation")), Occur.MUST)
+            .add(
+                SortedNumericDocValuesField.newSlowRangeQuery(
+                    "date_of_review", 20190210L, 20260210L),
+                Occur.FILTER)
+            .build();
+
+    WeightedSpanTermExtractor extractor = new WeightedSpanTermExtractor();
+    Map<String, WeightedSpanTerm> terms =
+        extractor.getWeightedSpanTerms(
+            query, 1, new CannedTokenStream(new Token("compensation", 0, 12)), "text");
+    assertTrue("Term 'compensation' should be extracted", terms.containsKey("compensation"));
   }
 
   @Override

--- a/lucene/highlighter/src/test/org/apache/lucene/search/highlight/custom/TestHighlightCustomQuery.java
+++ b/lucene/highlighter/src/test/org/apache/lucene/search/highlight/custom/TestHighlightCustomQuery.java
@@ -97,6 +97,31 @@ public class TestHighlightCustomQuery extends LuceneTestCase {
     assertEquals(Collections.singleton("quux"), terms.keySet());
   }
 
+  public void testHighlightQueryRequiringMetadata()
+      throws IOException, InvalidTokenOffsetsException {
+    String text = "The blue galaxy contains many bright stars";
+
+    // Anonymous subclass of the existing CustomQuery to inject the metadata call
+    CustomQuery q =
+        new CustomQuery(new Term(FIELD_NAME, "galaxy")) {
+          @Override
+          public Query rewrite(IndexSearcher indexSearcher) throws IOException {
+            if (indexSearcher.getIndexReader() instanceof org.apache.lucene.index.LeafReader) {
+              ((org.apache.lucene.index.LeafReader) indexSearcher.getIndexReader()).getFieldInfos();
+            }
+            return super.rewrite(indexSearcher);
+          }
+        };
+
+    String expected = "The blue <B>galaxy</B> contains many bright stars";
+    String observed = highlightField(q, FIELD_NAME, text);
+
+    assertEquals(
+        "Highlighter should support queries that perform schema/metadata lookups during rewrite",
+        expected,
+        observed);
+  }
+
   /** This method intended for use with <code>testHighlightingWithDefaultField()</code> */
   private String highlightField(Query query, String fieldName, String text)
       throws IOException, InvalidTokenOffsetsException {

